File: handwritten_ocr.py

### 1. `handwritten_ocr.py` (Python)

```python
# handwritten_ocr.py — Python версия

import sys
import os
import time
import argparse
import json
from pathlib import Path
from PIL import Image
import pytesseract
from colorama import init, Fore, Style

init(autoreset=True)

class HandwrittenOCR:
    def __init__(self, lang='eng', config=None):
        self.lang = lang
        self.config = config or '--psm 6 --oem 3'
        self.results = []

    def ocr_image(self, image_path):
        """Распознаёт рукописный текст с изображения."""
        try:
            img = Image.open(image_path)
            # Предобработка для улучшения распознавания
            gray = img.convert('L')
            text = pytesseract.image_to_string(
                gray,
                lang=self.lang,
                config=self.config
            )
            return text.strip()
        except Exception as e:
            return f"❌ Ошибка: {e}"

    def ocr_batch(self, image_dir):
        """Пакетная обработка изображений."""
        extensions = ('.jpg', '.jpeg', '.png', '.bmp', '.tiff', '.tif', '.webp')
        files = [f for f in Path(image_dir).iterdir() if f.suffix.lower() in extensions]

        if not files:
            print(Fore.YELLOW + "Нет поддерживаемых изображений.")
            return

        print(f"📁 Найдено {len(files)} изображений.")
        for i, file in enumerate(files, 1):
            print(f"\n[{i}/{len(files)}] Обработка: {file.name}")
            text = self.ocr_image(str(file))
            self.results.append({'file': file.name, 'text': text})
            self.save_text(text, file.stem + '.txt')

    def save_text(self, text, filename):
        with open(filename, 'w', encoding='utf-8') as f:
            f.write(text)
        print(f"💾 Сохранено: {filename}")

    def save_json(self, filename='ocr_results.json'):
        with open(filename, 'w', encoding='utf-8') as f:
            json.dump(self.results, f, indent=2, ensure_ascii=False)
        print(f"💾 Сохранено JSON: {filename}")

    def print_result(self, text, elapsed):
        print(f"\n{Fore.GREEN}Результат:{Style.RESET_ALL}")
        print("─" * 50)
        print(text)
        print("─" * 50)

        words = len(text.split())
        chars = len(text)
        print(f"{Fore.CYAN}📊 Статистика:{Style.RESET_ALL}")
        print(f"  Слов: {words}")
        print(f"  Символов: {chars}")
        print(f"  Время: {elapsed:.2f} сек")

def main():
    parser = argparse.ArgumentParser(description='Handwritten OCR Reader')
    parser.add_argument('input', help='Путь к изображению или папке')
    parser.add_argument('--lang', '-l', default='eng', help='Язык OCR (по умолчанию eng)')
    parser.add_argument('--psm', type=int, default=6, help='Режим сегментации (6 - блок текста)')
    parser.add_argument('--oem', type=int, default=3, help='Режим движка (3 - LSTM)')
    parser.add_argument('--batch', '-b', action='store_true', help='Пакетная обработка папки')
    parser.add_argument('--output', '-o', help='Файл для сохранения результата')
    args = parser.parse_args()

    print(Fore.CYAN + "✍️ OCR Handwritten Reader (Python)")

    config = f'--psm {args.psm} --oem {args.oem}'
    ocr = HandwrittenOCR(args.lang, config)

    if args.batch or os.path.isdir(args.input):
        ocr.ocr_batch(args.input)
        ocr.save_json()
    else:
        if not os.path.exists(args.input):
            print(Fore.RED + f"❌ Файл не найден: {args.input}")
            sys.exit(1)

        start = time.time()
        text = ocr.ocr_image(args.input)
        elapsed = time.time() - start

        ocr.print_result(text, elapsed)

        if args.output:
            ocr.save_text(text, args.output)
        else:
            default_output = os.path.splitext(args.input)[0] + '.txt'
            ocr.save_text(text, default_output)

if __name__ == "__main__":
    main()

File: handwritten_ocr.php
<?php
// handwritten_ocr.php — PHP версия

require_once 'vendor/autoload.php';

use thiagoalessio\TesseractOCR\TesseractOCR;

class HandwrittenOCR {
    private $lang;
    private $psm;
    private $oem;
    private $results = [];

    public function __construct($lang = 'eng', $psm = 6, $oem = 3) {
        $this->lang = $lang;
        $this->psm = $psm;
        $this->oem = $oem;
    }

    public function ocrImage($imagePath) {
        echo "📂 Обработка: " . basename($imagePath) . "\n";
        $start = microtime(true);

        try {
            $ocr = new TesseractOCR($imagePath);
            $ocr->lang($this->lang);
            $ocr->psm($this->psm);
            $ocr->oem($this->oem);
            $text = trim((string) $ocr);
            $elapsed = microtime(true) - $start;

            $this->printResult($text, $elapsed);
            $this->saveText($text, $imagePath);
            return $text;
        } catch (Exception $e) {
            echo "\033[31m❌ Ошибка: " . $e->getMessage() . "\033[0m\n";
            return null;
        }
    }

    public function ocrBatch($dir) {
        $exts = ['jpg', 'jpeg', 'png', 'bmp', 'tiff', 'webp'];
        $files = array_diff(scandir($dir), ['.', '..']);
        $images = [];
        foreach ($files as $f) {
            $ext = strtolower(pathinfo($f, PATHINFO_EXTENSION));
            if (in_array($ext, $exts)) {
                $images[] = $f;
            }
        }

        if (empty($images)) {
            echo "❌ Нет поддерживаемых изображений.\n";
            return;
        }

        echo "📁 Найдено " . count($images) . " изображений.\n";
        foreach ($images as $i => $f) {
            echo "\n[" . ($i+1) . "/" . count($images) . "] Обработка: $f\n";
            $text = $this->ocrImage($dir . DIRECTORY_SEPARATOR . $f);
            if ($text) {
                $this->results[] = ['file' => $f, 'text' => $text];
            }
        }

        $this->saveJSON('ocr_results.json');
    }

    private function printResult($text, $elapsed) {
        echo "\n\033[32mРезультат:\033[0m\n";
        echo str_repeat("─", 50) . "\n";
        echo $text . "\n";
        echo str_repeat("─", 50) . "\n";

        $words = count(array_filter(explode(' ', $text)));
        $chars = strlen($text);
        echo "\033[36m📊 Статистика:\033[0m\n";
        echo "  Слов: $words\n";
        echo "  Символов: $chars\n";
        echo "  Время: " . number_format($elapsed, 2) . " сек\n";
    }

    private function saveText($text, $imagePath) {
        $outFile = pathinfo($imagePath, PATHINFO_FILENAME) . '.txt';
        file_put_contents($outFile, $text);
        echo "💾 Сохранено: $outFile\n";
    }

    private function saveJSON($filename) {
        file_put_contents($filename, json_encode($this->results, JSON_PRETTY_PRINT | JSON_UNESCAPED_UNICODE));
        echo "💾 Сохранено JSON: $filename\n";
    }
}

function main($argv) {
    $input = null;
    $lang = 'eng';
    $batch = false;
    $output = null;

    for ($i = 1; $i < count($argv); $i++) {
        if ($argv[$i] == '--lang' || $argv[$i] == '-l') {
            $lang = $argv[++$i];
        } elseif ($argv[$i] == '--batch' || $argv[$i] == '-b') {
            $batch = true;
        } elseif ($argv[$i] == '--output' || $argv[$i] == '-o') {
            $output = $argv[++$i];
        } elseif ($input === null) {
            $input = $argv[$i];
        }
    }

    if ($input === null) {
        echo "Usage: php handwritten_ocr.php <image> [--lang eng] [--batch]\n";
        exit(1);
    }

    echo "\033[36m✍️ OCR Handwritten Reader (PHP)\033[0m\n";

    $ocr = new HandwrittenOCR($lang);

    if ($batch || is_dir($input)) {
        $ocr->ocrBatch($input);
    } else {
        if (!file_exists($input)) {
            echo "\033[31m❌ Файл не найден: $input\033[0m\n";
            exit(1);
        }
        $ocr->ocrImage($input);
    }
}

$argc = $_SERVER['argc'] ?? 0;
$argv = $_SERVER['argv'] ?? [];
main($argv);
?>
